File: clients/go/akeda/webhook_test.go
package akeda_test

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"testing"
	"time"

	"github.com/Artogceo/akeda-sdk/clients/go/akeda"
	"github.com/Artogceo/akeda-sdk/snapshot"
)

// Векторы берутся ИЗ СНИМКА, а не из копии в тесте. Копия — это вторая
// реализация правды: она переживает смену алгоритма и оставляет тест зелёным
// ровно тогда, когда он обязан покраснеть.
func contract(t *testing.T) snapshot.DeliveryContract {
	t.Helper()
	parsed, err := snapshot.ReadDeliveryContract()
	if err != nil {
		t.Fatalf("контракт доставки: %v", err)
	}
	if len(parsed.Vectors) < 2 {
		t.Fatalf("векторов должно быть больше одного, получено %d", len(parsed.Vectors))
	}
	return parsed
}

func headersFor(vector snapshot.Vector, signature string) http.Header {
	if signature == "" {
		signature = vector.Signature
	}
	headers := http.Header{}
	headers.Set(akeda.HeaderSignature, akeda.SignatureVersion+"="+signature)
	headers.Set(akeda.HeaderSignatureKeyID, vector.KeyID)
	headers.Set(akeda.HeaderSignatureTimestamp, strconv.FormatInt(vector.TimestampUnix, 10))
	headers.Set(akeda.HeaderInstallationID, vector.InstallationID)
	headers.Set(akeda.HeaderEventID, vector.EventID)
	return headers
}

func signatureOf(vector snapshot.Vector) akeda.Signature {
	return akeda.Signature{
		Version:        akeda.SignatureVersion,
		KeyID:          vector.KeyID,
		InstallationID: vector.InstallationID,
		EventID:        vector.EventID,
		IssuedAt:       time.Unix(vector.TimestampUnix, 0).UTC(),
	}
}

func TestConstantsMatchPublishedContract(t *testing.T) {
	parsed := contract(t)
	if akeda.SignatureVersion != parsed.Signature.Version {
		t.Fatalf("версия подписи: %s против %s", akeda.SignatureVersion, parsed.Signature.Version)
	}
	if int(akeda.DefaultSignatureWindow/time.Second) != parsed.Signature.WindowSeconds {
		t.Fatalf("окно: %v против %d секунд", akeda.DefaultSignatureWindow, parsed.Signature.WindowSeconds)
	}
	if parsed.Signature.Algorithm != "HMAC-SHA256" {
		t.Fatalf("алгоритм: %s", parsed.Signature.Algorithm)
	}
	want := map[string]string{
		"signature":      akeda.HeaderSignature,
		"keyId":          akeda.HeaderSignatureKeyID,
		"timestamp":      akeda.HeaderSignatureTimestamp,
		"installationId": akeda.HeaderInstallationID,
		"eventId":        akeda.HeaderEventID,
	}
	for name, value := range want {
		if parsed.Signature.Headers[name] != value {
			t.Fatalf("заголовок %s: %q против %q", name, parsed.Signature.Headers[name], value)
		}
	}
}

func TestSignatureMatchesContractVectors(t *testing.T) {
	for _, vector := range contract(t).Vectors {
		got := akeda.Sign(akeda.NewSigningKey(vector.KeyID, vector.Secret), signatureOf(vector), []byte(vector.Body))
		if got != vector.Signature {
			t.Fatalf("%s: подпись %s, ожидалась %s", vector.Name, got, vector.Signature)
		}
	}
}

func TestLegitimateDeliveryAccepted(t *testing.T) {
	parsed := contract(t)
	for _, vector := range parsed.Vectors {
		keys := []akeda.SigningKey{akeda.NewSigningKey(vector.KeyID, vector.Secret)}
		envelope, signature, err := akeda.VerifyWebhook(
			headersFor(vector, "").Get, []byte(vector.Body), keys,
			time.Unix(vector.TimestampUnix, 0), 0,
		)
		if err != nil {
			t.Fatalf("%s: %v", vector.Name, err)
		}
		if envelope.EventID != vector.EventID || signature.KeyID != vector.KeyID {
			t.Fatalf("%s: конверт не тот", vector.Name)
		}
	}
}

func TestTamperedBodyRejected(t *testing.T) {
	vector := contract(t).Vectors[0]
	tampered := strings.Replace(vector.Body, `"payload":`, `"injected":true,"payload":`, 1)
	_, _, err := akeda.VerifyWebhook(
		headersFor(vector, "").Get, []byte(tampered),
		[]akeda.SigningKey{akeda.NewSigningKey(vector.KeyID, vector.Secret)},
		time.Unix(vector.TimestampUnix, 0), 0,
	)
	if !errors.Is(err, akeda.ErrSignatureMismatch) {
		t.Fatalf("ожидался ErrSignatureMismatch, получено %v", err)
	}
}

func TestWrongSecretRejected(t *testing.T) {
	vector := contract(t).Vectors[0]
	_, _, err := akeda.VerifyWebhook(
		headersFor(vector, "").Get, []byte(vector.Body),
		[]akeda.SigningKey{akeda.NewSigningKey(vector.KeyID, "whs_"+strings.Repeat("0", 64))},
		time.Unix(vector.TimestampUnix, 0), 0,
	)
	if !errors.Is(err, akeda.ErrSignatureMismatch) {
		t.Fatalf("ожидался ErrSignatureMismatch, получено %v", err)
	}
}

func TestUnknownKeyHasItsOwnReason(t *testing.T) {
	vector := contract(t).Vectors[0]
	_, _, err := akeda.VerifyWebhook(
		headersFor(vector, "").Get, []byte(vector.Body),
		[]akeda.SigningKey{akeda.NewSigningKey("whk_other", vector.Secret)},
		time.Unix(vector.TimestampUnix, 0), 0,
	)
	if !errors.Is(err, akeda.ErrSignatureUnknownKey) {
		t.Fatalf("ожидался ErrSignatureUnknownKey, получено %v", err)
	}
}

func TestRotationAcceptsAnyActiveKey(t *testing.T) {
	vector := contract(t).Vectors[0]
	keys := []akeda.SigningKey{
		akeda.NewSigningKey("whk_new", "whs_"+strings.Repeat("a", 64)),
		akeda.NewSigningKey(vector.KeyID, vector.Secret),
	}
	if _, _, err := akeda.VerifyWebhook(
		headersFor(vector, "").Get, []byte(vector.Body), keys,
		time.Unix(vector.TimestampUnix, 0), 0,
	); err != nil {
		t.Fatalf("подпись предыдущим ключом обязана проходить перекрытие ротации: %v", err)
	}
}

func TestMissingHeadersRejected(t *testing.T) {
	vector := contract(t).Vectors[0]
	_, _, err := akeda.VerifyWebhook(http.Header{}.Get, []byte(vector.Body), nil, time.Now(), 0)
	if !errors.Is(err, akeda.ErrSignatureMissing) {
		t.Fatalf("ожидался ErrSignatureMissing, получено %v", err)
	}
}

func TestUnknownVersionRejected(t *testing.T) {
	vector := contract(t).Vectors[0]
	headers := headersFor(vector, "")
	headers.Set(akeda.HeaderSignature, "v2="+vector.Signature)
	_, _, err := akeda.VerifyWebhook(
		headers.Get, []byte(vector.Body),
		[]akeda.SigningKey{akeda.NewSigningKey(vector.KeyID, vector.Secret)},
		time.Unix(vector.TimestampUnix, 0), 0,
	)
	if !errors.Is(err, akeda.ErrSignatureVersionUnknown) {
		t.Fatalf("ожидался ErrSignatureVersionUnknown, получено %v", err)
	}
}

func TestWindowIsTwoSided(t *testing.T) {
	parsed := contract(t)
	vector := parsed.Vectors[0]
	keys := []akeda.SigningKey{akeda.NewSigningKey(vector.KeyID, vector.Secret)}
	beyond := time.Duration(parsed.Signature.WindowSeconds+60) * time.Second
	for _, shift := range []time.Duration{beyond, -beyond} {
		_, _, err := akeda.VerifyWebhook(
			headersFor(vector, "").Get, []byte(vector.Body), keys,
			time.Unix(vector.TimestampUnix, 0).Add(shift), 0,
		)
		if !errors.Is(err, akeda.ErrSignatureExpired) {
			t.Fatalf("смещение %v: ожидался ErrSignatureExpired, получено %v", shift, err)
		}
	}
	edge := time.Duration(parsed.Signature.WindowSeconds) * time.Second
	if _, _, err := akeda.VerifyWebhook(
		headersFor(vector, "").Get, []byte(vector.Body), keys,
		time.Unix(vector.TimestampUnix, 0).Add(edge), 0,
	); err != nil {
		t.Fatalf("на границе окна доставка ещё принимается: %v", err)
	}
}

func TestHeaderBodyMismatchRejected(t *testing.T) {
	parsed := contract(t)
	vector, other := parsed.Vectors[0], parsed.Vectors[1]
	key := akeda.NewSigningKey(vector.KeyID, vector.Secret)
	// Подписываем ЧУЖОЙ event_id в заголовке своим ключом: подпись сойдётся,
	// потому что заголовок входит в подписываемую строку целиком.
	forged := signatureOf(vector)
	forged.EventID = other.EventID
	value := akeda.Sign(key, forged, []byte(vector.Body))
	headers := headersFor(vector, value)
	headers.Set(akeda.HeaderEventID, other.EventID)
	_, _, err := akeda.VerifyWebhook(
		headers.Get, []byte(vector.Body), []akeda.SigningKey{key},
		time.Unix(vector.TimestampUnix, 0), 0,
	)
	if !errors.Is(err, akeda.ErrEventMismatch) {
		t.Fatalf("ожидался ErrEventMismatch, получено %v", err)
	}
}

func TestEnvelopeWithoutRequiredFieldRejected(t *testing.T) {
	vector := contract(t).Vectors[0]
	key := akeda.NewSigningKey(vector.KeyID, vector.Secret)
	var raw map[string]any
	if err := json.Unmarshal([]byte(vector.Body), &raw); err != nil {
		t.Fatalf("вектор не разбирается: %v", err)
	}
	raw["trace_id"] = ""
	body, err := json.Marshal(raw)
	if err != nil {
		t.Fatalf("сборка тела: %v", err)
	}
	value := akeda.Sign(key, signatureOf(vector), body)
	_, _, err = akeda.VerifyWebhook(
		headersFor(vector, value).Get, body, []akeda.SigningKey{key},
		time.Unix(vector.TimestampUnix, 0), 0,
	)
	if !errors.Is(err, akeda.ErrEnvelopeInvalid) {
		t.Fatalf("ожидался ErrEnvelopeInvalid, получено %v", err)
	}
}

func TestSecretNeverPrinted(t *testing.T) {
	vector := contract(t).Vectors[0]
	key := akeda.NewSigningKey(vector.KeyID, vector.Secret)
	for _, printed := range []string{key.String(), key.GoString()} {
		if strings.Contains(printed, vector.Secret) {
			t.Fatalf("секрет попал в печать ключа: %s", printed)
		}
	}
	encoded, err := json.Marshal(key)
	if err != nil {
		t.Fatalf("маршалинг ключа: %v", err)
	}
	if strings.Contains(string(encoded), vector.Secret) {
		t.Fatalf("секрет попал в JSON ключа: %s", encoded)
	}
}

File: clients/typescript/test/webhook.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { deliveryContract } from "./contract.js";
import {
  DEFAULT_WINDOW_MS,
  SIGNATURE_VERSION,
  WebhookVerificationError,
  parseSignature,
  sign,
  verifyWebhook,
} from "../src/webhook.js";

function headersFor(vector: (typeof deliveryContract.vectors)[number], signature = vector.signature) {
  return {
    "Akeda-Signature": `${SIGNATURE_VERSION}=${signature}`,
    "Akeda-Signature-Key-Id": vector.keyId,
    "Akeda-Signature-Timestamp": String(vector.timestampUnix),
    "Akeda-Installation-Id": vector.installationId,
    "Akeda-Event-Id": vector.eventId,
  };
}

function at(vector: (typeof deliveryContract.vectors)[number], offsetSeconds = 0): Date {
  return new Date((vector.timestampUnix + offsetSeconds) * 1000);
}

test("константы совпадают с опубликованным контрактом", () => {
  assert.equal(SIGNATURE_VERSION, deliveryContract.signature.version);
  assert.equal(DEFAULT_WINDOW_MS, deliveryContract.signature.windowSeconds * 1000);
  assert.equal(deliveryContract.signature.algorithm, "HMAC-SHA256");
});

test("подпись сходится с векторами контракта", () => {
  assert.ok(deliveryContract.vectors.length >= 2, "векторов должно быть больше одного");
  for (const vector of deliveryContract.vectors) {
    const produced = sign(
      { id: vector.keyId, secret: vector.secret },
      {
        keyId: vector.keyId,
        installationId: vector.installationId,
        eventId: vector.eventId,
        issuedAtUnix: vector.timestampUnix,
      },
      vector.body,
    );
    assert.equal(produced, vector.signature, vector.name);
  }
});

test("законная доставка принимается и конверт разбирается", () => {
  for (const vector of deliveryContract.vectors) {
    const { envelope, signature } = verifyWebhook(headersFor(vector), vector.body, {
      keys: { id: vector.keyId, secret: vector.secret },
      now: at(vector),
    });
    assert.equal(envelope.event_id, vector.eventId);
    assert.equal(envelope.installation_id, vector.installationId);
    assert.equal(signature.keyId, vector.keyId);
    for (const field of deliveryContract.envelope.requiredFields) {
      assert.ok(field in envelope, `в конверте нет ${field}`);
    }
  }
});

test("подмена тела после подписи отвергается", () => {
  const vector = deliveryContract.vectors[0]!;
  const tampered = vector.body.replace('"payload":', '"injected":true,"payload":');
  assert.throws(
    () => verifyWebhook(headersFor(vector), tampered, {
      keys: { id: vector.keyId, secret: vector.secret },
      now: at(vector),
    }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "mismatch",
  );
});

test("чужой секрет при знакомом идентификаторе ключа отвергается", () => {
  const vector = deliveryContract.vectors[0]!;
  assert.throws(
    () => verifyWebhook(headersFor(vector), vector.body, {
      keys: { id: vector.keyId, secret: "whs_" + "0".repeat(64) },
      now: at(vector),
    }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "mismatch",
  );
});

test("незнакомый идентификатор ключа отвергается отдельной причиной", () => {
  const vector = deliveryContract.vectors[0]!;
  assert.throws(
    () => verifyWebhook(headersFor(vector), vector.body, {
      keys: { id: "whk_другой", secret: vector.secret },
      now: at(vector),
    }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "unknown_key",
  );
});

test("ротация: подходит любой из действующих ключей", () => {
  const vector = deliveryContract.vectors[0]!;
  const { envelope } = verifyWebhook(headersFor(vector), vector.body, {
    keys: [
      { id: "whk_новый", secret: "whs_" + "a".repeat(64) },
      { id: vector.keyId, secret: vector.secret },
    ],
    now: at(vector),
  });
  assert.equal(envelope.event_id, vector.eventId);
});

test("доставка без заголовков подписи отвергается", () => {
  const vector = deliveryContract.vectors[0]!;
  assert.throws(
    () => verifyWebhook({}, vector.body, { keys: { id: vector.keyId, secret: vector.secret } }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "missing",
  );
});

test("неизвестная версия схемы подписи отвергается", () => {
  const vector = deliveryContract.vectors[0]!;
  const headers = { ...headersFor(vector), "Akeda-Signature": `v2=${vector.signature}` };
  assert.throws(
    () => verifyWebhook(headers, vector.body, {
      keys: { id: vector.keyId, secret: vector.secret },
      now: at(vector),
    }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "version_unknown",
  );
});

test("окно двустороннее: и просроченная подпись, и подпись из будущего", () => {
  const vector = deliveryContract.vectors[0]!;
  const keys = { id: vector.keyId, secret: vector.secret };
  const beyond = deliveryContract.signature.windowSeconds + 60;
  for (const offset of [beyond, -beyond]) {
    assert.throws(
      () => verifyWebhook(headersFor(vector), vector.body, { keys, now: at(vector, offset) }),
      (error: unknown) => error instanceof WebhookVerificationError && error.reason === "expired",
      `смещение ${offset}`,
    );
  }
  // На границе окна доставка ещё принимается.
  verifyWebhook(headersFor(vector), vector.body, {
    keys,
    now: at(vector, deliveryContract.signature.windowSeconds),
  });
});

test("заголовки про одно событие, тело про другое — отказ", () => {
  const vector = deliveryContract.vectors[0]!;
  const other = deliveryContract.vectors[1]!;
  // Подписываем ЧУЖОЙ event_id в заголовке своим ключом: подпись сойдётся,
  // потому что заголовок входит в подписываемую строку целиком.
  const parts = {
    keyId: vector.keyId,
    installationId: vector.installationId,
    eventId: other.eventId,
    issuedAtUnix: vector.timestampUnix,
  };
  const signature = sign({ id: vector.keyId, secret: vector.secret }, parts, vector.body);
  const headers = { ...headersFor(vector, signature), "Akeda-Event-Id": other.eventId };
  assert.throws(
    () => verifyWebhook(headers, vector.body, {
      keys: { id: vector.keyId, secret: vector.secret },
      now: at(vector),
    }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "event_mismatch",
  );
});

test("конверт без обязательного поля отвергается", () => {
  const vector = deliveryContract.vectors[0]!;
  const body = JSON.stringify({ ...JSON.parse(vector.body), trace_id: "" });
  const parts = {
    keyId: vector.keyId,
    installationId: vector.installationId,
    eventId: vector.eventId,
    issuedAtUnix: vector.timestampUnix,
  };
  const signature = sign({ id: vector.keyId, secret: vector.secret }, parts, body);
  assert.throws(
    () => verifyWebhook(headersFor(vector, signature), body, {
      keys: { id: vector.keyId, secret: vector.secret },
      now: at(vector),
    }),
    (error: unknown) => error instanceof WebhookVerificationError && error.reason === "envelope_invalid",
  );
});

test("заголовки читаются без учёта регистра", () => {
  const vector = deliveryContract.vectors[0]!;
  const lowered = Object.fromEntries(
    Object.entries(headersFor(vector)).map(([name, value]) => [name.toLowerCase(), value]),
  );
  const signature = parseSignature((name) => lowered[name.toLowerCase()] ?? null);
  assert.equal(signature.eventId, vector.eventId);
});
